x-pack/plugins/observability_solution/observability/server/services/index.ts

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import * as t from 'io-ts';
+import {
+  IScopedClusterClient,
+  IUiSettingsClient,
+  KibanaRequest,
+  SavedObjectsClientContract,
+} from '@kbn/core/server';
+import { LicensingApiRequestHandlerContext } from '@kbn/licensing-plugin/server';
+import { concat } from 'lodash';
+
+export const observabilityAlertDetailsContextRt = t.intersection([
+  t.type({
+    alert_started_at: t.string,
+  }),
+  t.partial({
+    // apm fields
+    'service.name': t.string,
+    'service.environment': t.string,
+    'transaction.type': t.string,
+    'transaction.name': t.string,
+
+    // infrastructure fields
+    'host.name': t.string,
+    'container.id': t.string,
+    'kubernetes.pod.name': t.string,
+  }),
+]);
+
+export type AlertDetailsContextualInsightsHandlerQuery = t.TypeOf<
+  typeof observabilityAlertDetailsContextRt
+>;
+
+export interface AlertDetailsContextualInsight {
+  key: string;
+  description: string;
+  data: unknown;
+}
+
+export interface AlertDetailsContextualInsightsRequestContext {
+  request: KibanaRequest;
+  core: Promise<{
+    elasticsearch: {
+      client: IScopedClusterClient;
+    };
+    uiSettings: {
+      client: IUiSettingsClient;
+      globalClient: IUiSettingsClient;
+    };
+    savedObjects: {
+      client: SavedObjectsClientContract;
+    };
+  }>;
+  licensing: Promise<LicensingApiRequestHandlerContext>;
+}
+type AlertDetailsContextualInsightsHandler = (
+  context: AlertDetailsContextualInsightsRequestContext,
+  query: AlertDetailsContextualInsightsHandlerQuery
+) => Promise<AlertDetailsContextualInsight[]>;
+
+export class AlertDetailsContextualInsightsService {
+  private handlers: AlertDetailsContextualInsightsHandler[] = [];
+
+  constructor() {}
+
+  registerHandler(handler: AlertDetailsContextualInsightsHandler) {
+    this.handlers.push(handler);
+  }
+
+  async getAlertDetailsContext(
+    context: AlertDetailsContextualInsightsRequestContext,
```

Comment:
Similar to `initResources` I am not sure it's a good idea to re-invent a new "god object" that we pass around. I'd prefer if we can stick to the base arguments that Kibana provides and that we need downstream.

Reply:
Actually, thinking about this again we probably shouldn't pass `core` and `plugins`. That should be available to the consumer already through their own plugin. What we need to pass them are the stateful things like a scoped ES client and the request (specifically the query params).